File: ts/project/.sdk/tm/scala/core/Context.scala
package SCALAPACKAGE.core

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.concurrent.ThreadLocalRandom
import SCALAPACKAGE.utility.struct.Struct

// Per-operation context threaded through the pipeline and feature hooks.
class Context(ctxmap: JMap[String, Object], basectx: Context) {

  var id: String = "C" + (ThreadLocalRandom.current().nextInt(90000000) + 10000000)
  var out: JMap[String, Object] = new LinkedHashMap[String, Object]()
  var ctrl: Control = new Control()
  var meta: JMap[String, Object] = new LinkedHashMap[String, Object]()
  var client: SdkClient = null
  var utility: Utility = null
  var op: Operation = null
  var point: JMap[String, Object] = null
  var config: JMap[String, Object] = null
  var entopts: JMap[String, Object] = null
  var options: JMap[String, Object] = null
  var opmap: JMap[String, Operation] = null
  var response: Response = null
  var result: Result = null
  var spec: Spec = null
  var data: JMap[String, Object] = null
  var reqdata: JMap[String, Object] = null
  var matchData: JMap[String, Object] = null
  var reqmatch: JMap[String, Object] = null
  var entity: Entity = null
  var shared: JMap[String, Object] = null

  locally {
    // Client
    Helpers.getCtxProp(ctxmap, "client") match { case c: SdkClient => client = c; case _ => }
    if (client == null && basectx != null) client = basectx.client

    // Utility
    Helpers.getCtxProp(ctxmap, "utility") match { case u: Utility => utility = u; case _ => }
    if (utility == null && basectx != null) utility = basectx.utility

    // Ctrl
    val cv = Helpers.getCtxProp(ctxmap, "ctrl")
    if (cv != null) {
      cv match {
        case cm: JMap[_, _] =>
          val m = cm.asInstanceOf[JMap[String, Object]]
          m.get("throw") match { case t: java.lang.Boolean => ctrl.throwing = t; case _ => }
          m.get("explain") match { case e: JMap[_, _] => ctrl.explain = e.asInstanceOf[JMap[String, Object]]; case _ => }
          m.get("actor") match { case a: String => ctrl.actor = a; case _ => }
          m.get("paging") match { case p: JMap[_, _] => ctrl.paging = p.asInstanceOf[JMap[String, Object]]; case _ => }
        case c: Control => ctrl = c
        case _ =>
      }
    } else if (basectx != null && basectx.ctrl != null) {
      ctrl = basectx.ctrl
    }

    // Meta
    Helpers.getCtxProp(ctxmap, "meta") match {
      case m: JMap[_, _] => meta = m.asInstanceOf[JMap[String, Object]]
      case _ => if (basectx != null && basectx.meta != null) meta = basectx.meta
    }

    // Config
    Helpers.getCtxProp(ctxmap, "config") match { case cfg: JMap[_, _] => config = cfg.asInstanceOf[JMap[String, Object]]; case _ => }
    if (config == null && basectx != null) config = basectx.config

    // Entopts
    Helpers.getCtxProp(ctxmap, "entopts") match { case eo: JMap[_, _] => entopts = eo.asInstanceOf[JMap[String, Object]]; case _ => }
    if (entopts == null && basectx != null) entopts = basectx.entopts

    // Options
    Helpers.getCtxProp(ctxmap, "options") match { case o: JMap[_, _] => options = o.asInstanceOf[JMap[String, Object]]; case _ => }
    if (options == null && basectx != null) options = basectx.options

    // Entity
    Helpers.getCtxProp(ctxmap, "entity") match { case e: Entity => entity = e; case _ => }
    if (entity == null && basectx != null) entity = basectx.entity

    // Shared
    Helpers.getCtxProp(ctxmap, "shared") match { case s: JMap[_, _] => shared = s.asInstanceOf[JMap[String, Object]]; case _ => }
    if (shared == null && basectx != null) shared = basectx.shared

    // Opmap
    Helpers.getCtxProp(ctxmap, "opmap") match { case om: JMap[_, _] => opmap = om.asInstanceOf[JMap[String, Operation]]; case _ => }
    if (opmap == null && basectx != null) opmap = basectx.opmap
    if (opmap == null) opmap = new LinkedHashMap[String, Operation]()

    // Data
    data = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "data"))
    if (data == null) data = new LinkedHashMap[String, Object]()
    reqdata = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqdata"))
    if (reqdata == null) reqdata = new LinkedHashMap[String, Object]()
    matchData = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "match"))
    if (matchData == null) matchData = new LinkedHashMap[String, Object]()
    reqmatch = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqmatch"))
    if (reqmatch == null) reqmatch = new LinkedHashMap[String, Object]()

    // Point
    Helpers.getCtxProp(ctxmap, "point") match { case t: JMap[_, _] => point = t.asInstanceOf[JMap[String, Object]]; case _ => }
    if (point == null && basectx != null) point = basectx.point

    // Spec
    Helpers.getCtxProp(ctxmap, "spec") match { case sp: Spec => spec = sp; case _ => }
    if (spec == null && basectx != null) spec = basectx.spec

    // Result
    Helpers.getCtxProp(ctxmap, "result") match { case r: Result => result = r; case _ => }
    if (result == null && basectx != null) result = basectx.result

    // Response
    Helpers.getCtxProp(ctxmap, "response") match { case resp: Response => response = resp; case _ => }
    if (response == null && basectx != null) response = basectx.response

    // Resolve operation
    val opname = Helpers.getCtxProp(ctxmap, "opname") match { case s: String => s; case _ => "" }
    op = resolveOp(opname)
  }

  private def resolveOp(opname: String): Operation = {
    // Cache key is `<entity>:<opname>` so two entities with the same op get
    // distinct cached Operations.
    var entname = ""
    if (entity != null) entname = entity.getName()
    val cacheKey = entname + ":" + opname

    val cached = opmap.get(cacheKey)
    if (cached != null) return cached

    if ("" == opname) return new Operation(new LinkedHashMap[String, Object]())

    val opcfg = Struct.getpath(config, java.util.List.of("entity", entname, "op", opname))

    var input = "match"
    if ("update" == opname || "create" == opname) input = "data"

    var points: JList[Object] = null
    opcfg match {
      case _: JMap[_, _] =>
        Struct.getprop(opcfg, "points") match {
          case t: JList[_] => points = t.asInstanceOf[JList[Object]]
          case _ =>
        }
      case _ =>
    }
    if (points == null) points = new ArrayList[Object]()

    val opdef = new LinkedHashMap[String, Object]()
    opdef.put("entity", entname)
    opdef.put("name", opname)
    opdef.put("input", input)
    opdef.put("points", points)

    val newop = new Operation(opdef)
    opmap.put(cacheKey, newop)
    newop
  }

  def makeError(code: String, msg: String): SdkError = new SdkError(code, msg, this)
}

File: ts/project/.sdk/tm/java/core/Context.java
package JAVAPACKAGE.core;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;

import JAVAPACKAGE.utility.struct.Struct;

/** Per-operation context threaded through the pipeline and feature hooks. */
@SuppressWarnings({"unchecked"})
public class Context {

  public String id;
  public Map<String, Object> out = new LinkedHashMap<>();
  public Control ctrl;
  public Map<String, Object> meta;
  public SdkClient client;
  public Utility utility;
  public Operation op;
  public Map<String, Object> point;
  public Map<String, Object> config;
  public Map<String, Object> entopts;
  public Map<String, Object> options;
  public Map<String, Operation> opmap;
  public Response response;
  public Result result;
  public Spec spec;
  public Map<String, Object> data;
  public Map<String, Object> reqdata;
  public Map<String, Object> match;
  public Map<String, Object> reqmatch;
  public Entity entity;
  public Map<String, Object> shared;

  public Context(Map<String, Object> ctxmap, Context basectx) {
    this.id = "C" + (ThreadLocalRandom.current().nextInt(90000000) + 10000000);

    // Client
    Object c = Helpers.getCtxProp(ctxmap, "client");
    if (c instanceof SdkClient) {
      this.client = (SdkClient) c;
    }
    if (this.client == null && basectx != null) {
      this.client = basectx.client;
    }

    // Utility
    Object u = Helpers.getCtxProp(ctxmap, "utility");
    if (u instanceof Utility) {
      this.utility = (Utility) u;
    }
    if (this.utility == null && basectx != null) {
      this.utility = basectx.utility;
    }

    // Ctrl
    this.ctrl = new Control();
    Object cv = Helpers.getCtxProp(ctxmap, "ctrl");
    if (cv != null) {
      if (cv instanceof Map) {
        Map<String, Object> cm = (Map<String, Object>) cv;
        Object t = cm.get("throw");
        if (t instanceof Boolean) {
          this.ctrl.throwing = (Boolean) t;
        }
        Object e = cm.get("explain");
        if (e instanceof Map) {
          this.ctrl.explain = (Map<String, Object>) e;
        }
        Object a = cm.get("actor");
        if (a instanceof String) {
          this.ctrl.actor = (String) a;
        }
        Object p = cm.get("paging");
        if (p instanceof Map) {
          this.ctrl.paging = (Map<String, Object>) p;
        }
      }
      else if (cv instanceof Control) {
        this.ctrl = (Control) cv;
      }
    }
    else if (basectx != null && basectx.ctrl != null) {
      this.ctrl = basectx.ctrl;
    }

    // Meta
    this.meta = new LinkedHashMap<>();
    Object m = Helpers.getCtxProp(ctxmap, "meta");
    if (m instanceof Map) {
      this.meta = (Map<String, Object>) m;
    }
    else if (basectx != null && basectx.meta != null) {
      this.meta = basectx.meta;
    }

    // Config
    Object cfg = Helpers.getCtxProp(ctxmap, "config");
    if (cfg instanceof Map) {
      this.config = (Map<String, Object>) cfg;
    }
    if (this.config == null && basectx != null) {
      this.config = basectx.config;
    }

    // Entopts
    Object eo = Helpers.getCtxProp(ctxmap, "entopts");
    if (eo instanceof Map) {
      this.entopts = (Map<String, Object>) eo;
    }
    if (this.entopts == null && basectx != null) {
      this.entopts = basectx.entopts;
    }

    // Options
    Object o = Helpers.getCtxProp(ctxmap, "options");
    if (o instanceof Map) {
      this.options = (Map<String, Object>) o;
    }
    if (this.options == null && basectx != null) {
      this.options = basectx.options;
    }

    // Entity
    Object e = Helpers.getCtxProp(ctxmap, "entity");
    if (e instanceof Entity) {
      this.entity = (Entity) e;
    }
    if (this.entity == null && basectx != null) {
      this.entity = basectx.entity;
    }

    // Shared
    Object s = Helpers.getCtxProp(ctxmap, "shared");
    if (s instanceof Map) {
      this.shared = (Map<String, Object>) s;
    }
    if (this.shared == null && basectx != null) {
      this.shared = basectx.shared;
    }

    // Opmap
    Object om = Helpers.getCtxProp(ctxmap, "opmap");
    if (om instanceof Map) {
      this.opmap = (Map<String, Operation>) om;
    }
    if (this.opmap == null && basectx != null) {
      this.opmap = basectx.opmap;
    }
    if (this.opmap == null) {
      this.opmap = new LinkedHashMap<>();
    }

    // Data
    this.data = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "data"));
    if (this.data == null) {
      this.data = new LinkedHashMap<>();
    }
    this.reqdata = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqdata"));
    if (this.reqdata == null) {
      this.reqdata = new LinkedHashMap<>();
    }
    this.match = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "match"));
    if (this.match == null) {
      this.match = new LinkedHashMap<>();
    }
    this.reqmatch = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqmatch"));
    if (this.reqmatch == null) {
      this.reqmatch = new LinkedHashMap<>();
    }

    // Point
    Object t = Helpers.getCtxProp(ctxmap, "point");
    if (t instanceof Map) {
      this.point = (Map<String, Object>) t;
    }
    if (this.point == null && basectx != null) {
      this.point = basectx.point;
    }

    // Spec
    Object sp = Helpers.getCtxProp(ctxmap, "spec");
    if (sp instanceof Spec) {
      this.spec = (Spec) sp;
    }
    if (this.spec == null && basectx != null) {
      this.spec = basectx.spec;
    }

    // Result
    Object r = Helpers.getCtxProp(ctxmap, "result");
    if (r instanceof Result) {
      this.result = (Result) r;
    }
    if (this.result == null && basectx != null) {
      this.result = basectx.result;
    }

    // Response
    Object resp = Helpers.getCtxProp(ctxmap, "response");
    if (resp instanceof Response) {
      this.response = (Response) resp;
    }
    if (this.response == null && basectx != null) {
      this.response = basectx.response;
    }

    // Resolve operation
    Object opname = Helpers.getCtxProp(ctxmap, "opname");
    this.op = resolveOp(opname instanceof String ? (String) opname : "");
  }

  private Operation resolveOp(String opname) {
    // Cache key is `<entity>:<opname>` so two entities with the same op
    // (e.g. both have a "list") get distinct cached Operations. Keying on
    // opname alone caused the first-resolved entity's points to be served
    // to every subsequent entity's call.
    String entname = "";
    if (this.entity != null) {
      entname = this.entity.getName();
    }
    String cacheKey = entname + ":" + opname;

    Operation cached = this.opmap.get(cacheKey);
    if (cached != null) {
      return cached;
    }

    if ("".equals(opname)) {
      return new Operation(new LinkedHashMap<>());
    }

    Object opcfg = Struct.getpath(this.config,
        List.of("entity", entname, "op", opname));

    String input = "match";
    if ("update".equals(opname) || "create".equals(opname)) {
      input = "data";
    }

    List<Object> points = null;
    if (opcfg instanceof Map) {
      Object t = Struct.getprop(opcfg, "points");
      if (t instanceof List) {
        points = (List<Object>) t;
      }
    }
    if (points == null) {
      points = new ArrayList<>();
    }

    Map<String, Object> opdef = new LinkedHashMap<>();
    opdef.put("entity", entname);
    opdef.put("name", opname);
    opdef.put("input", input);
    opdef.put("points", points);

    Operation op = new Operation(opdef);

    this.opmap.put(cacheKey, op);
    return op;
  }

  public SdkError makeError(String code, String msg) {
    return new SdkError(code, msg, this);
  }
}
